### ISSUE TITLE
macOS MIDI: memory leak when freeing song

### ISSUE THREAD

**Issue by SDLBugzilla** [libsdl-org/SDL_mixer#258]
# This bug report was migrated from our old Bugzilla tracker.

**Reported in version:** 2.0.4
**Reported for operating system, platform:** macOS 10.13, x86_64

# Comments on the original bug report:

On 2019-03-30 14:11:03 +0000, ioanch wrote:

> I've discovered a memory leak when using MIDI music on macOS (I'm on Mojave).
> 
> When freeing a song, DisposeMusicSequence returns error "kAudioToolboxErr_CannotDoInCurrentContext" (-10863, https://www.osstatus.com/search/results?platform=all&framework=all&search=-10863 ), unless you call "MusicPlayerSetSequence(song->player, NULL);" before it.
> 
> See my fix I did on my fork of SDL2_mixer here: https://bitbucket.org/ichera/sdl_mixer/commits/713acbd976efd31c441e37906d76b7fa790572dd?at=macos-music-leak-fix

On 2019-04-04 09:39:04 +0000, Ozkan Sezer wrote:

> Ryan, is this patch correct?

On 2019-07-03 08:32:13 +0000, Ozkan Sezer wrote:

> Ryan:  Should we apply this?

On 2019-07-03 09:32:16 +0000, Sam Lantinga wrote:

> Patch added, thanks!
> https://hg.libsdl.org/SDL_mixer/rev/1c557e8ac98f

**Comment by sezero** [comment]
A segmentation fault was reported to me with this patch on El Capitan in
the SDL-1.2 branch. @ioan-chera, @slouken, @icculus: Can anyone reproduce?
Can anyone reproduce with main SDL2 branch?
```
El Cap 10.11.6
Xcode Version 8.1 (8B62)

$gcc --version
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/usr/include/c++/4.2.1
Apple LLVM version 8.0.0 (clang-800.0.42.1)
Target: x86_64-apple-darwin15.6.0
```
```
Playing /Users/admin/new/ABBA_-_Dancing_Queen.mid
Segmentation fault: 11 (core dumped)

$lldb --core /cores/core.53424
(lldb) target create --core "/cores/core.53424"
warning: (x86_64) /cores/core.53424 load command 111 LC_SEGMENT_64 has a fileoff + filesize (0x28ff2000) that extends beyond the end of the file (0x28ff1000), the segment will be truncated to match
warning: (x86_64) /cores/core.53424 load command 112 LC_SEGMENT_64 has a fileoff (0x28ff2000) that extends beyond the end of the file (0x28ff1000), ignoring this section
Core file '/cores/core.53424' (x86_64) was loaded.
(lldb) bt
* thread #1: tid = 0x0000, 0x00007fff93bc98ea libsystem_kernel.dylib`__kill + 10, stop reason = signal SIGSTOP
  * frame #0: 0x00007fff93bc98ea libsystem_kernel.dylib`__kill + 10
    frame #1: 0x00007fff926f852a libsystem_platform.dylib`_sigtramp + 26
    frame #2: 0x00007fff8d6de84a AudioToolbox`Sequence::LazyCreateDefaultGraph() + 10
    frame #3: 0x00007fff8d6e0d3c AudioToolbox`SequencePlayer::CleanupPlayState() + 32
    frame #4: 0x00007fff8d6e0a82 AudioToolbox`SequencePlayer::~SequencePlayer() + 40
    frame #5: 0x00007fff8d6e0dec AudioToolbox`SequencePlayer::~SequencePlayer() + 14
    frame #6: 0x00007fff8d6daeb9 AudioToolbox`DisposeMusicPlayer + 27
    frame #7: 0x00000001052321f5 libSDL_mixer-1.2.0.dylib`native_midi_freesong(song=0x00007fe183d2f480) + 69 at native_midi_macosx.c:261 [opt]
    frame #8: 0x0000000105230e58 libSDL_mixer-1.2.0.dylib`Mix_FreeMusic(music=<unavailable>) + 168 at music.c:789 [opt]
    frame #9: 0x000000010519cb2c playmus`main(argc=<unavailable>, argv=<unavailable>) + 1020 at playmus.c:233 [opt]
    frame #10: 0x00007fff96f6f5ad libdyld.dylib`start + 1
    frame #11: 0x00007fff96f6f5ad libdyld.dylib`start + 1
```

**Comment by sezero** [comment]
PING for https://github.com/libsdl-org/SDL_mixer/issues/258#issuecomment-931260899 ?

**Comment by sezero** [comment]
Cherry-picked the fix to SDL-1.2 branch as https://github.com/libsdl-org/SDL_mixer/commit/3bc1b796e53a45a8df944a8dcf7ccb6a48a7681a
